### PR TITLE
[Ready][MDB IGNORE] Bot Multi-Z Movement

### LINKED ISSUE #732: Little bots can't patrol
## Round ID:
Multiple rounds

## Reproduction:
Build any of the little bots, be it a medibot, cleanbot, firebot, etc. Set it to patrol mode, it says "Engaging patrol mode", then  immediately "Disengaging patrol mode" and does not patrol, albeit it moves and does its job if the jobs target is close enough

### LINKED ISSUE #1973: Bot Pathfinding
## Testmerges:
Issue discovered due to the test merge of [#1956: 'AI Bot Control Fix' by Bobbanz1](https://github.com/BeeStation/NSV13/pull/1956)

## Reproduction:
Bots are unable to calculate a route, for pathfinding, in and out of a room which has a door currently blocked with directional fire doors. 
Issue is temporarily solved through manually forcing every bot to find a way into a nearby room that isn't being blocked by firelocks? through maintenance, which also has chance to return a failed attempt at pathfinding. 

Once the bots have found their way to this room then one can manually open the door and assuming the air alarms doesn't trigger, the bots can navigate through this open door into this previously blocked off area, however if the firelock closed infront of the bot then, assuming it still has the route in memory, it will continue through the firelocks and to its destination. 

However if the route wasn't calculate before the firelocks closed down, then it will result in the pathfinding attempt failing. 

The bots also seem confused by elevators such as on the Aetherwhisp, sometimes they fail to calculate a route out of the elevator despite there being nothing that's blocking their way. 

Sometimes they also fail to calculate a route in the elevator itself even if they're standing inside the elevator.

### PR DESCRIPTION
**Pull request by Bobbanz1**
## About The Pull Request
Right, first of all!
Resolves #1973
Resolves #732

Edit: 
Ports the a modified version of the PR from upstream: [BeeStation/BeeStation-Hornet/pull/7587](https://github.com/BeeStation/BeeStation-Hornet/pull/7587)
Also dependant on PR: [BeeStation/NSV13/pull/2050](https://github.com/BeeStation/NSV13/pull/2050)

**Information the Mappers need to know**
Bot Elevators are placed down in the same manner that Ladders are placed down, IE, on the same tile coordinates on different Z-levels. 

Navigation Beacons are now linked to the overmap ship itself, this means two things: 

- The Bots will now go through the entire list of all beacons on the map, whenever they’re looking for their next beacon, potentially slowing them down if there are too many beacons. 
- In order to form a network with a beacon on another level the beacon closest to the elevator will need to have their destination be a beacon on the next Z-level, like Deck1Up with the destination being Deck2Receive. 
- - Now the beacon telling them to go upstairs doesn’t have to be directly next to the elevator as that was an example, however the pathfinding will try to find the fastest path to get to the elevator and the search might not cover the entire ship before the pathfinding gives up, so test if they can find a path to the elevator.

**Back to your regular schedule program**

While this PR is capable of being used on any ship that has a bot elevator on it, if the AI is the one doing the summoning, however it can't do the Multi-Z Patrol thing just yet, as that requires all ships to have their navigation beacons reprogrammed from going in one big loop to instead going from destination to destination and then finally up a ladder, after which the next deck will also need a route that has a navigation beacon programmed to make the bot go downwards. 

(This can easily be implemented by copying over the location name of a beacon downstairs, as the next_patrol destination of some navbeacon upstairs)

Ports some of the code introduced by this PR: [tgstation/tgstation/pull/65906](https://github.com/tgstation/tgstation/pull/65906)
Full credit for the code used to find the closest ladder goes to cacogen.
Although the code has been modified

- [x] Implements the ability for bots to go between z-levels
- [x] Implement the ability for bots to patrol through z-levels
- [x] Implement the ability to summon bots from other z-levels
- [x] Implement the ability for the AI to call bots to other z-levels  
- [x] Implement this ability for Mulebots to deliver stuff between z-levels
- [x] Implement Patrol Network for Eclipse
- [x] Implement multi-Z patrol network to all the currently played maps
- - [x] Atlas
- - [x] Aetherwhisp
- - [x] Tycoon
- - [x] Galactica
- - [x] Snake
- [x] Implement Mulebot Delivery areas on all the maps where it makes sense for them
- - [x] Atlas
- - [x] Aetherwhisp
- - [x] Tycoon
- - [x] Galactica
- - [x] Snake

Second of all!
Some bots, possibly all of them, are now capable of moving between Z-levels if one of the following conditions are met:

- The AI has slammed down a waypoint on another Z-level, like on the bottom deck or the upper deck.
- - If this happens then the bot will instantly head for the nearest bot elevator it can find from its current position, upon arriving to this bot elevator it will check the original waypoint and use the bot elevator to either go up or down, afterwards it resets the bot call command before promptly running it yet again, this time with the original waypoint as the final destination.
- - - This annoyingly produces a text message in the chat saying that the call command has been reset.
- A Navigation Beacon has a destination that's located on another Z-level, currently it's restricted to only check Z-level 2 and 3 which seems to always be the Z-levels the ships various decks are located on. 
- - Upon detecting the beacons location on another deck, the bot will try heading for the nearest bot elevator possible.
- - I've introduced a new structure which uses the same code as the ladders however this thing is specifically designed to be used by bots and will reject anyone else trying to use it to go up or down, this thing is called a bot elevator, the reason I introduced this thing is because...well it's easier for mappers to just replicate what they do with ladders for these things and then no longer have to worry about it.
- - - Upon reaching the bot elevator it will go down, or up it, and promptly forget the original destination and start heading for the nearest beacon it can find (could possibly also be trying to reach the next beacon from the original destination one)
- A PDA has Summoned this bot.

Third of all!
The Bot Elevators
They're currently using a temporary upside down version of the bin icon from the mining folder and act the same as ladders except for only being usable by bots, these things are created because *cough*_Moisturizer_*cough* wanted ladders for bots only. 

Anyways for any future mappers reading this, basically just do the same thing you do with ladders for these things.

Currently only the Atlas has a somewhat semi-working patrol network, although it sometimes gets stuck upstairs if the bot originally started its route on the navigation beacon outside the Head of Security's office. 

And now for the horrible, _horrible_ reveal, these bots are using KNPC Z-level travel to be capable of doing this stuff.

## Why It's Good For The Game
Bot uses bot elevators. Bot is now advanced boi capable of patrolling.

## Changelog
:cl: 
add: Added the ability for bots to go up and down bot elevators.
add: Added the ability for bots to find a route to another Z-level if called to it.
add: Added the ability for bots to have a patrol route going downstairs or upstairs using bot elevators.
add: Added the ability for bots to be summoned using a PDA on another z-level.
add: Added the ability for mulebots to travel between Z-levels.
add: Gave Medibots Maintenance Access, this is an NSV13 only thing as some maps have bot elevators in maintenance areas.
add: Added some Mulebots to various maps for multi-Z delivery testing.
add: Added the ability to summon bots from other station Z-levels in the bot control computer program
/:cl:

### CODE REVIEW

**Review by Bokkiewokkie — APPROVED** [review]
Mapping approved

File: code/__DEFINES/access.dm [hunk 1]
```diff
@@ -75,6 +75,8 @@
 #define ACCESS_PRISONER 78 //! Nsv13 - Genpop. This access is added to prisoner IDs that have served their time.
 #define ACCESS_HANGAR 79 //NSV13 - Access to the hangar
 
+
+
```
[the remark this answers is not in the document]
Comment (Bobbanz1) [reply code/__DEFINES/access.dm:79]: Seems to be leftover from when I had the idea of trying to implement an access specifically for bots.

[the remark this answers is not in the document]
Comment (Bobbanz1) [reply code/__DEFINES/access.dm]: It has been resolved

File: code/modules/mob/living/simple_animal/bot/medbot.dm [hunk 2]
```diff
@@ -92,6 +92,7 @@ GLOBAL_VAR(medibot_unique_id_gen)
 	. = ..()
 	var/datum/job/doctor/J = new /datum/job/doctor
 	access_card.access += J.get_access()
+	access_card.access += ACCESS_MAINT_TUNNELS //NSV13 - DIFFERENCE BETWEEN CODEBASE
```
Comment (Bokkiewokkie) [inline code/modules/mob/living/simple_animal/bot/medbot.dm:95]:
```suggestion
	access_card.access += ACCESS_MAINT_TUNNELS //NSV13 - Added maintenance access for bots
```
Please describe the type of change made in the comments.

**Review by Bokkiewokkie — CHANGES_REQUESTED** [review]

**Review by Bokkiewokkie — APPROVED** [review]

**Pull request merged**